### DISCUSSION TITLE
Connecting the `ThanosRuler` CRD with an existing Thanos + prometheus-operator setup

### DISCUSSION THREAD

**Discussion by parkedwards** [prometheus-operator/prometheus-operator#4612] [Q&A] [answered]
Hey there, 

I have an existing cluster that implements the prometheus-operator via the [bitnami kube-prometheus chart](https://github.com/bitnami/charts/tree/master/bitnami/kube-prometheus/) -- I realize that this chart is largely out of scope for discussion in this repo.  

However, I'm just looking to add the `ThanosRuler` component to my cluster via the CRD defined in this repository/project, and move some rule evaluations to this component.

I've been able to provision the ruler component via the `ThanosRuler` CRD, pull in rules via existing `PrometheusRules` custom resources via `ruleSelector`s ✅ (I see the ruler component evaluating these rules in the pod logs, as well as seeing the rulefiles in the container's fs)...

**but I'm having some trouble understanding how to get this Ruler to properly speak to the rest** of my Thanos/Prometheus components.  

As a reference, I already had existing, functional instances of:
* Prometheus nodes
* Alertmanager instances
* Thanos store, storegateway, query, query-frontend, and sidecars on the Prometheus-es

To illustrate, I'm not seeing any alerting rules (green or red) in my Thanos UI, which tells me that the Ruler is trying (but failing) to push these rule evals all around.  I looked through the `thanos-ruler` examples in this repo, but having some trouble pinpointing some config args that might help me solve this.  

Could anyone help point me in the right direction?

<img width="1192" alt="image" src="https://user-images.githubusercontent.com/19556538/156674626-32846891-aaed-470c-b054-64c53da4b80a.png">

For reference, here is the set of `rule` args that my ThanosRuler CRD pod is running against

```Containers:
  thanos-ruler:
    Container ID:  <some-hash>
    Image:         <my-repo>/thanos/thanos:v0.23.0
    Ports:         10901/TCP, 10902/TCP
    Host Ports:    0/TCP, 0/TCP
    Args:
      rule
      --data-dir=/thanos/data
      --eval-interval=15s
      --tsdb.retention=24h
      --label=thanos_ruler_replica="$(POD_NAME)"
      --alert.label-drop=thanos_ruler_replica
      --log.level=warn
      --query=dnssrv+_http._tcp.thanos-query.monitoring.svc.cluster.local
      --rule-file=/etc/thanos/rules/*/*.yaml
      --alertmanagers.url=http://prometheus-kube-prometheus-alertmanager.monitoring:9093
      --objstore.config=$(OBJSTORE_CONFIG)
```

**Comment by parkedwards** [comment]
one thing i'm noticing in other Ruler implementations is that there seem to be these args passed into the `thanos rule` exec:

```
        - --grpc-address=0.0.0.0:10901
        - --http-address=0.0.0.0:10902
```

but i'm not seeing these as pass-able args into the `ThanosRulerSpec` schema
https://github.com/prometheus-operator/prometheus-operator/blob/main/Documentation/api.md#thanosrulerspec

**Comment by metalmatze** [comment, marked as answer]
The Thanos ruler only speaks to a query endpoint that you configured in 
```
      --query=dnssrv+_http._tcp.thanos-query.monitoring.svc.cluster.local
```
With that the Ruler will be able to find and query the Thanos Querier correctly. Eventually the new data (blocks) will be send to object storage.
If you want find the Thanos Ruler's data until the data gets uploaded you need to configure the Thanos Querier to find the Thanos Ruler, just like the Querier needs to be configured to find the Thanos Store.
Here are the flags on my own Thanos Querier that should help you figure out what you need:
```
      query
      --store=dnssrv+_grpc._tcp.thanos-store.monitoring.svc.cluster.local
      --store=dnssrv+_grpc._tcp.prometheus-operated.monitoring.svc.cluster.local
      --store=dnssrv+_grpc._tcp.thanos-ruler-metalmatze.monitoring.svc.cluster.local
      ...
```

**Reply by parkedwards** [reply]
ah this is very helpful. thanks @metalmatze !
